flow/src/lib.rs

```diff
@@ -0,0 +1,975 @@
+		UnsuccessfulCampaign,
+
+		//
+		//	create
+		//
+		/// Campaign must end after it starts
+		EndTooEarly,
+		/// Campaign expiry has be lower than the block number limit
+		EndTooLate,
+		/// Max contributions per block exceeded
+		ContributionsPerBlockExceeded,
+		/// Name too long
+		NameTooLong,
+		/// Name too short
+		NameTooShort,
+		/// Deposit exceeds the campaign target
+		DepositTooHigh,
+		/// Campaign id exists
+		IdExists,
+
+		//
+		//	mint
+		//
+		/// Overflow adding a new campaign to total fundings
+		AddCampaignOverflow,
+		/// Overflow adding a new owner
+		AddOwnedOverflow,
+		/// Overflow adding to the total number of contributors of a camapaign
+		UpdateContributorOverflow,
+		/// Overflow adding to the total number of contributions of a camapaign
+		AddContributionOverflow,
+		/// Campaign owner unknown
+		OwnerUnknown,
+		/// Campaign admin unknown
+		AdminUnknown,
+		/// Cannot contribute to owned campaign
+		NoContributionToOwnCampaign,
+		/// Guru Meditation
+		GuruMeditation,
+		/// Zou are not authorized for this call
+		AuthorizationError,
+		/// Contributions not allowed
+		NoContributionsAllowed,
+		/// Id Unknown
+		IdUnknown,
+		/// Transfer Error
+		TransferError,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// Block finalization
+		fn on_finalize(_n: BlockNumberFor<T>) {
+			// get all the campaigns ending in current block
+			let block_number = <frame_system::Pallet<T>>::block_number();
+			// which campaigns end in this block
+			let campaign_hashes = Self::campaigns_by_block(block_number);
+
+			// iterate over campaigns ending in this block
+			for campaign_id in &campaign_hashes {
+				// get campaign struct
+				let campaign = Self::campaign_by_id(campaign_id);
+				let campaign_balance = Self::campaign_balance(campaign_id);
+				let dao = Self::campaign_org(&campaign_id);
+				let dao_treasury = T::Control::body_treasury(&dao);
+
+				// check for cap reached
+				if campaign_balance >= campaign.cap {
+					// get campaign owner
+					// should be controller --- test?
+					let _owner = Self::campaign_owner(campaign_id);
+
+					match _owner {
+						Some(owner) => {
+							// get all contributors
+							let contributors = Self::campaign_contributors(campaign_id);
+							let mut transaction_complete = true;
+
+							// 1 iterate over contributors
+							// 2 unreserve contribution
+							// 3 transfer contribution to campaign treasury
+							'inner: for contributor in &contributors {
+								// if contributor == campaign owner, skip
+								if contributor == &owner {
+									continue;
+								}
+
+								// get amount from contributor
+								let contributor_balance = Self::campaign_contribution((
+									*campaign_id,
+									contributor.clone(),
+								));
+
+								// unreserve the amount in contributor balance
+								T::Currency::unreserve(
+									T::FundingCurrencyId::get(),
```

Comment:
You can store result of `T::FundingCurrencyId::get()` to some variable and use it instead, since this value is used in multiple places over the `on_finalize` hook.

Reply:
Thank you. Both getters refactoring, and this issue will be handled later during the hardening stage.